### PR TITLE
Add policy backend router

### LINKED ISSUE #2611: Add support for configuring policy backend
Allow configuring a (single) backend for getting a PD for a scope and for checking the access policy.
If no backend is configured, it must fall back to the file-based PD store. The authorized endpoint for that store always returns true. This could result in an unsafe config, better to force use through `strictmode` (disallow, PD-files only) 

requires #2606

### PR DESCRIPTION
**Pull request by woutslakhorst**
closes #2611 

~disclaimer: changes are two weeks old, can't remember what the state is. :this_is_fine:~

### CODE REVIEW

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: policy/policy.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+			return fmt.Errorf("failed to parse policy.address: %w", err)
+		}
+		tlsConfig, err := b.pkiInstance.CreateTLSConfig(config.TLS)
+		if err != nil {
+			return err
+		}
+		b.backend = &remote{
+			address: b.config.Address,
+			// todo client timeout
+			client: client.NewHTTPClient(config.Strictmode, 30*time.Second, tlsConfig),
```
Comment (reinkrul) [inline policy/policy.go:70]: 30s feels awfully long.

File: auth/api/iam/api.go [hunk 2]
```diff
@@ -335,18 +338,25 @@ func (r Wrapper) OAuthClientMetadata(ctx context.Context, request OAuthClientMet
...
 
+	authorizer, err := did.ParseDID(request.Did)
+	if err != nil {
+		return nil, oauth.OAuth2Error{
+			Code:        oauth.InvalidRequest,
+			Description: err.Error(),
+		}
+	}
+
 	// todo: only const scopes supported, scopes with variable arguments not supported yet
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply auth/api/iam/api.go:354]: doesn't matter, is now the responsibility of the backend.

File: policy/local.go [hunk 3]
```diff
@@ -16,25 +16,74 @@
...
+	mapping, ok := b.mapping[scope]
+	if !ok {
+		return nil, ErrNotFound
+	}
+	result := pe.PresentationDefinition(mapping)
+	return &result, nil
+}
+
+func (b *localStore) Authorized(_ context.Context, _ client.AuthorizedRequest) (bool, error) {
+	return true, nil
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply policy/local.go:51]: Yes, that's part of the responsibility of the authorization server (openID4VP)

File: policy/local.go [hunk 4]
```diff
@@ -16,25 +16,74 @@
...
+
+	// load all the files
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+		if !strings.HasSuffix(file.Name(), ".json") {
+			continue
+		}
+		err := s.loadFromFile(fmt.Sprintf("%s/%s", directory, file.Name()))
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply policy/local.go:77]: That would be the same class as errors when you misdefine values in the definition or wrong DIDs. That should be tested after deployment to development/test.

File: policy/interface.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+// ModuleName is the name of the policy module
+const ModuleName = "policy"
+
+var ErrNotFound = errors.New("not found")
+
+// Backend is the interface for the policy backend
+// Both the remote and local policy backend implement this interface
+type Backend interface {
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply policy/interface.go:36]: Renamed to PDPBackend

File: policy/local.go [hunk 6]
```diff
@@ -16,25 +16,74 @@
...
 	"io"
 	"os"
+	"strings"
 )
 
-// DefinitionResolver is a store for presentation definitions
+// localStore is a backend for presentation definitions
 // It loads a file with the mapping from oauth scope to presentation definition
-type DefinitionResolver struct {
+type localStore struct {
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply policy/local.go:36]: renamed to localPDP

File: policy/policy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply policy/policy.go:70]: type of auth will be determined based on feedback of first users.

Comment (woutslakhorst) [reply policy/policy.go:70]: replaced with client http timeout config

File: auth/api/iam/api.go [hunk 7]
```diff
@@ -335,18 +338,24 @@ func (r Wrapper) OAuthClientMetadata(ctx context.Context, request OAuthClientMet
...
 	if len(request.Params.Scope) == 0 {
 		return PresentationDefinition200JSONResponse(PresentationDefinition{}), nil
 	}
 
-	// todo: only const scopes supported, scopes with variable arguments not supported yet
-	// todo: we only take the first scope as main scope, when backends are introduced we need to use all scopes and send them as one to the backend.
-	scopes := strings.Split(request.Params.Scope, " ")
-	presentationDefinition := r.auth.PresentationDefinitions().ByScope(scopes[0])
-	if presentationDefinition == nil {
+	authorizer, err := did.ParseDID(request.Did)
```
Comment (reinkrul) [inline auth/api/iam/api.go:346]: should we check whether the node "owns" the DID, to potentially unburden a REST API backend that queries the XIS' database?

Comment (woutslakhorst) [reply auth/api/iam/api.go:346]: 'Normal' operation would not be optimised, just unsanctioned access. A generic security assessment on the 'open' APIs would be a better approach. (IsOwned goes to Vault...)

**Review by reinkrul — APPROVED** [review]

**Pull request merged**